source/docs/quiche_integration.md

```diff
@@ -33,5 +33,7 @@ QUICHE doesn't buffer data at the local connection layer. All the data is buffer
 When the aggregated buffered bytes goes above high watermark, its registered network callbacks will call Network::ConnectionCallbacks::onAboveWriteBufferHighWatermark(). The HCM will notify each stream via QUIC codec Http::Connection::onUnderlyingConnectionAboveWriteBufferHighWatermark() which will call each stream's StreamCallbackHelper::runHighWatermarkCallbacks(). There might be a way to simply the call stack as Quic connection already knows about all the stream, there is no need to call to HCM and notify each stream via codec. But here we just follow the same logic as HTTP2 codec does. In the same way, any QuicStream::OnCanWrite() may change the aggregated buffered bytes in the connection level bookkeeping as well. If the buffered bytes goes down below the low watermark, the same calls will be triggered to propagate onBelowWriteBufferLowWatermark() to each stream.
 
 As to Http::StreamEncoder::encodeHeaders()/encodeTrailers(), the accounting is done differently between Google QUIC and IETF QUIC:
- * In Google QUIC, encodeHeaders()/encodeTrailers() check the buffer size increase on header stream before and after writing headers/trailers. In QuicSession::OnCanWrite(), may drain header stream send buffer, so there we also check send buffer size decrease on header stream.
+ * In Google QUIC, encodeHeaders()/encodeTrailers() check the buffer size increase on header stream before and after writing headers/trailers. The bytes buffered in headers stream doesn't contribute to the data stream's send buffer watermark, but contributes to the connection's send buffer watermark. QuicSession::OnCanWrite() may drain header stream send buffer, so there we also check send buffer size decrease on header stream. Because there is no way to track which bytes in headers stream belongs to which data stream, if we had counted the send buffer increase in headers stream against the data stream on which encodeHeaders()/encodeTrailers() were called, there would be no way to discount these bytes when they were drained in OnCanWrite().
  * In IETF QUIC, encodeHeaders()/encodeTrailers() check the buffer size increase on the corresponding data stream which is similar to encodeData(). The buffered headers/trailers are only drained via QuicStream::OnCanWrite() so there is no need to check QuicSession::OnCanWrite.
+
+When the stream gets closed due to error, the send buffer might not be empty. In such case, these bytes are discounted from the send buffer watermarks of stream and connection to allow other streams to send.
```

Comment:
high level comment, one of the OOM vectors that Envoy had with HTTP, was an incoming badly formed request generating a 400 error response.  If the client kept sending malformed requests without reading from the TCP socket, the error responses would queue up in the server side send buffer, and Envoy would eventually OOM.
If we are discounting bytes, is there some upper bound where we'll simply close the connection (and if so, maybe expand this documentation), or is it opening up an OOM vector which we need to address?

Reply:
In the case you described, the stream buffers the 400 response and FIN, and it won't get closed. The connection is bound by max stream per connection limit. 
I believe H2 also has similar upper bound for active stream. I'm curious why the H2 client can still send more request?

In addition to that, once a quic stream gets closed, it will be destroyed after all the sent bytes are ACKed or, if it is reset, will be destroyed immediately. The send buffer will be destroyed at that time. 
The code here is handling the stream reset case, otherwise the stream only gets closed after all the bytes in the send buffer have been sent.

If the client keeps open stream after receiving 400 response but doesn't ACK those responses. QUIC connection can buffer unlimited number of closed streams waiting for ACK, but the total number of unACKed bytes are bound by congestion control. In that sense, the number of closed streams are bounded too. QUICHE also has mechanism to close connection if there are too many outstanding packets.

Reply:
That sounds like we're pretty solid.  Would you be up for adding this to the docs?  Beyond addressing my concerns, I think it's a fantastic explanation to have in the code base for others who want to understand how we guard against OOM vectors here :-)

Reply:
done